ToolBox/PsychSignalDataConverter/PsychSignalDataDownloader.cs

```diff
@@ -0,0 +1,170 @@
+
+        /// <summary>
+        /// Downlods data from psychsignal
+        /// </summary>
+        /// <param name="rawDataDestination">Directory we write raw data to</param>
+        /// <param name="apiKey">PsychSignal API key</param>
+        /// <param name="dataSource">Data source (e.g. stocktwits,twitter_withretweets)</param>
+        public PsychSignalDataDownloader(string rawDataDestination, string apiKey, string dataSource)
+        {
+            _rawDataDestination = rawDataDestination;
+
+            _dataSource = dataSource;
+            _apiKey = apiKey;
+        }
+
+        /// <summary>
+        /// Download the data from the given starting date to the ending date.
+        /// Note that if the ending date is in the same hour as the current time,
+        /// we will lower the <paramref name="endDateUtc" /> by one hour in order
+        /// to make sure that we only download complete, non-changing dataa
+        /// </summary>
+        /// <param name="startDateUtc">Starting date</param>
+        /// <param name="endDateUtc">Ending date</param>
+        public void Download(DateTime startDateUtc, DateTime endDateUtc)
+        {
+            var now = DateTime.UtcNow;
+            var nowHour = new DateTime(now.Year, now.Month, now.Day, now.Hour, 0, 0);
+
+            Directory.CreateDirectory(_rawDataDestination);
+
+            if (startDateUtc < now.AddDays(-15))
+            {
+                throw new ArgumentException("The starting date can only be at most 15 days from now");
+            }
+
+            // Makes sure we only get final, non-changing data by checking if the end date is greater than
+            // or equal to the current time and setting it to an hour before the current time if the condition is met
+            if (nowHour <= new DateTime(endDateUtc.Year, endDateUtc.Month, endDateUtc.Day, endDateUtc.Hour, 0, 0))
+            {
+                Log.Trace($"PsychSignalDataDownloader.Download(): Adjusting end time from {endDateUtc:yyyy-MM-dd HH:mm:ss} to {nowHour.AddHours(-1):yyyy-MM-dd HH:mm:ss}");
+                endDateUtc = nowHour.AddHours(-1);
+            }
+
+            // Get the total amount of hours in order to keep track of progress
+            var totalHours = endDateUtc.Subtract(startDateUtc).TotalHours;
+            var percentage = 1 / totalHours;
+            Log.Trace("PsychSignalDataDownloader.Download(): Begin downloading raw data");
+
+            // PsychSignal paginates data by hour. Note that it is possible to retrieve non-complete data if the requested hour
+            // is the same as the current hour or greater than the current hour.
+            for (var i = 1; startDateUtc < endDateUtc; startDateUtc = startDateUtc.AddHours(1), i++)
+            {
+                var rawDataPath = Path.Combine(_rawDataDestination, $"{startDateUtc:yyyyMMdd_HH}.csv");
+                var rawDataPathTemp = Path.Combine(Path.GetTempPath(), $"{Guid.NewGuid().ToString()}.csv.tmp");
```

Comment:
fyi, when using string interpolation, you don't need to call `ToString()`, so this is equally written as:
``` csharp
$"{Guid.NewGuid()}.csv.tmp"
```
And if you want to pass a format specifier, such as `N` to remove dashes, or when stringifying a date time, `O` to use ISO8601, you can use:
``` csharp
"${Guid.NewGuid():N}.csv.tmp"
```